File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        n = len(s)
        
        @cache
        def foo(i, bal):
            if i == n:
                return bal == 0
            if bal < 0:
                return False
            
            ans = False
            
            if s[i] == "(":
                ans = foo(i+1, bal+1)
            elif s[i] == ")":
                ans = foo(i+1, bal-1)
            else:
                ans = foo(i+1, bal) or foo(i+1, bal+1) or foo(i+1, bal-1)
                
            return ans
        
        return foo(0,0)

File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.java
class Solution {
    public boolean checkValidString(String s) {
        int l = 0;
        int h = 0;
        
        for (char i : s.toCharArray()){
            if (i == '('){
                l++; h++;
            }
            else if (i == ')'){
                l--; h--;
            }
            else {
                l--; h++;
            }
            if (h < 0){
                return false;
            }
            if (l < 0){
                l = 0;
            }
        }
        return l == 0;
    }
}